### ISSUE TITLE
Algorithms for surface detection

### ISSUE THREAD

**Issue by fuulish** [fuulish/surf#5]
Currently the code does a brute force search for voxels that belong to the surface. If one is found the location of the surface is determined from a linear interpolation within that voxel. It works and appears to be sufficiently accurate. Nevertheless, one could do better.

A nicer way could be to actually minimize g(x) - c = 0, where g(x) is coarse-grained density field (or equivalently search for a minimum in (g(x) -c)^2). This can be done after the brute force search as refinement step. 
One could also determine one point on the surface and walk along a path connecting neighboring surface voxels. I am not sure how to do the latter in full 3D mode and find all relevant points, but if it's clear that the surface is roughly in the (x,y)-plane then it's straightforward (or (x,z)/(y,z) plane for that matter).

**Comment by OndrejMarsalek** [comment]
I thought about this a bit and I don't think we should reinvent the wheel. There must be pretty standard CG algorithms to represent isosurfaces. I think we should look for those first. I think rough+refinement is generally a good idea, though.

**Comment by OndrejMarsalek** [comment]
I imagine, for example, that it should be possible to get a rough surface using a (relatively) coarse 3D grid and the interpolate/smoothen that, getting more points than would correspond to the 3D grid. Refinement could still go on top of that, ideally with something that maintains roughly equal density and spacing of points on that surface. I am sure CG has standard tools for that, but I don't know them.

**Comment by fuulish** [comment]
I agree, reinventing the wheel is definitely not the way to go. My suggestion would have been to keep the coarse brute-force search, because it manages to quickly capture the whole volume, and then do a refinement. For the latter I wanted to use a standard minimization algorithm from the GSL. That, I think, should not walk away from the initial guess too far. Here, restraining to a certain search direction would for example guarantee that it stays equally spaced within the other two directions without accuracy loss.

**Comment by OndrejMarsalek** [comment]
Yes, except that now limits you to specific geometries, i.e. a slab in a given direction, in this case. I think it's fine to just refine the regular points that you have, as a first pass, though it would be nice to eventually have something entirely robust. Eventually...
